src/ChannelServer/World/SkillHandler.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Melia.Channel.Network;
+using Melia.Shared.Data.Database;
+using Melia.Shared.World;
+
+namespace Melia.Channel.World
+{
+	interface ITargetedSkillHandler
+	{
+		void Handle(Skill skill, Character caster, IEntity target);
+	}
+
+	interface IGroundSkillHandler
+	{
+		void Handle(Skill skill, Character caster, Position position);
+	}
+
+	interface ITargetGroundSkillHandler
+	{
+		void Handle(Skill skill, Character caster, Position position, List<IEntity> targets);
+	}
+
+	class TargetedSkillHandler : ITargetedSkillHandler
+	{
+		public void Handle(Skill skill, Character caster, IEntity target)
+		{
+			if (skill.Data.OverHeat != 0)
+				skill.OverHeatCount++;
+			Send.ZC_OVERHEAT_CHANGED(caster, skill);
+			if (skill.Data.OverHeat == 0 || skill.Data.OverHeat == skill.OverHeatCount)
+			{
+				Send.ZC_COOLDOWN_CHANGED(caster, skill);
+				skill.OverHeatCount = 0;
+			}
+
+			var damage = caster.GetRandomPAtk() + 100;
+			Send.ZC_SKILL_FORCE_TARGET(caster, target, skill, damage);
+			// Unsafe cast for now we're only dealing with monsters
+			if (target != null)
+			{
+				var monster = (Monster)target;
+				if (monster.TakeDamage(damage, caster, 0))
+					Send.ZC_SKILL_CAST_CANCEL(caster, target);
+			}
+		}
+	}
+
+	class TargetedGroundSkillHandler : ITargetGroundSkillHandler
+	{
+		public void Handle(Skill skill, Character caster, Position position, List<IEntity> targets)
+		{
+			if (skill.Data.OverHeat != 0)
+				skill.OverHeatCount++;
+			Send.ZC_OVERHEAT_CHANGED(caster, skill);
+			if (skill.Data.OverHeat == 0 || skill.Data.OverHeat == skill.OverHeatCount)
+			{
+				Send.ZC_COOLDOWN_CHANGED(caster, skill);
+				skill.OverHeatCount = 0;
+			}
+			var damage = caster.GetRandomPAtk() + 100;
+			Send.ZC_SKILL_MELEE_GROUND(caster, skill, position.X, position.Y, position.Z, null, 0);
+			foreach (var target in targets)
+			{
+				//Send.ZC_SKILL_MELEE_GROUND(caster, skill, target.Position.X, target.Position.Y, target.Position.Z, null, 0);
+				var monster = (Monster)target;
+				if (monster.TakeDamage(damage, caster, 2))
+					Send.ZC_SKILL_CAST_CANCEL(caster, target);
+			}
+		}
+	}
+
+	class GroundSkillHandler : IGroundSkillHandler
+	{
+		public void Handle(Skill skill, Character caster, Position position)
+		{
+			if (skill.Data.OverHeat != 0)
+				skill.OverHeatCount++;
+			Send.ZC_OVERHEAT_CHANGED(caster, skill);
+			if (skill.Data.OverHeat == 0 || skill.Data.OverHeat == skill.OverHeatCount)
+			{
+				Send.ZC_COOLDOWN_CHANGED(caster, skill);
+				skill.OverHeatCount = 0;
+			}
+			Send.ZC_SKILL_READY(caster, skill, caster.Position, position);
+			switch (skill.Id)
+			{
+				case 20007:
+					caster.Map.GetVisibleMonsters(caster);
+					var targets = caster.Map.GetAttackableMonstersInRange(position.X, position.Y, position.Z, (int)skill.Data.MaxRange).ToList<IEntity>();
```

Comment:
It seems like a waste to allocate an entire new list, just to cast the list of monsters to a list of entities, especially since you don't seem to do anything that would require a list. Wouldn't it be more efficient to leave it as a monster list and make the type of `targets` on `ZC_SKILL_MELEE_GROUND` and `IEnumerable`? That way you don't need this allocation. Though if you were to add the interface I proposed for targets, you could return that from `GetAttackable` and use it without casts.